### PR TITLE
Optimize grouped conv bwd weight for small M and N

### PR DESCRIPTION
**Pull request by bartekxk**
- Add support for universal gemm for two stage grouped conv bwd wei
- Add support for merging multiple batches to M and N dimensions
- Tune instances

### CODE REVIEW

File: library/src/tensor_operation_instance/gpu/grouped_conv3d_bwd_weight/CMakeLists.txt [hunk 1]
```diff
@@ -1,36 +1,38 @@
...
+     wmma/device_grouped_conv3d_bwd_weight_wmma_ndhwgc_gkzyxc_ndhwgk_i8_instance.cpp)
 
-if((DTYPES MATCHES "fp8" AND DTYPES MATCHES "bf8" AND DTYPES MATCHES "fp16") OR NOT DEFINED DTYPES)
-    list(APPEND GROUPED_CONV3D_BWD_WEIGHT
-      xdl/device_grouped_conv3d_bwd_weight_xdl_ndhwgc_gkzyxc_ndhwgk_f16_comp_bf8_fp8_instance.cpp)
-endif()
+ if((DTYPES MATCHES "fp8" AND DTYPES MATCHES "bf8" AND DTYPES MATCHES "fp16") OR NOT DEFINED DTYPES)
+     list(APPEND GROUPED_CONV3D_BWD_WEIGHT
+       xdl/device_grouped_conv3d_bwd_weight_xdl_ndhwgc_gkzyxc_ndhwgk_f16_comp_bf8_fp8_instance.cpp)
+ endif()
```
Comment (aosewski) [inline library/src/tensor_operation_instance/gpu/grouped_conv3d_bwd_weight/CMakeLists.txt]: Those added whitespaces confuse what has changed. I think they're not needed.

File: library/src/tensor_operation_instance/gpu/grouped_conv2d_bwd_weight/xdl/device_grouped_conv2d_bwd_weight_two_stage_xdl_nhwgc_gkyxc_nhwgk_f16_pipev2_instance.cpp [hunk 2]
```diff
@@ -30,16 +30,9 @@ void add_device_grouped_conv2d_bwd_weight_two_stage_xdl_nhwgc_gkyxc_nhwgk_f16_in
...
-            ConvBwdWeightDefault>{});
-    // 2. Filter1x1Stride1Pad0
-    add_device_operation_instances(
-        instances,
-        device_grouped_conv_bwd_weight_two_stage_xdl_c_shuffle_f16_instances<
-            2,
-            NHWGC,
-            GKYXC,
-            NHWGK,
-            ConvBwdWeightFilter1x1Stride1Pad0>{});
```
Comment (aosewski) [inline library/src/tensor_operation_instance/gpu/grouped_conv2d_bwd_weight/xdl/device_grouped_conv2d_bwd_weight_two_stage_xdl_nhwgc_gkyxc_nhwgk_f16_pipev2_instance.cpp:42]: Is it not needed anymore?

File: include/ck/tensor_operation/operator_transform/transform_conv_bwd_weight_to_gemm_v2.hpp [hunk 3]
```diff
@@ -0,0 +1,640 @@
+
```
Comment (aosewski) [inline include/ck/tensor_operation/operator_transform/transform_conv_bwd_weight_to_gemm_v2.hpp]: Empty line.

File: include/ck/tensor_operation/operator_transform/transform_conv_bwd_weight_to_gemm_v2.hpp [hunk 4]
```diff
@@ -0,0 +1,640 @@
...
+namespace tensor_operation {
+
+/**
+ * @brief Transform conv bwd weight to gemm v2
+ *
+ * This version does following things:
+ * 1. Merge KBatch with K0 to align descriptor with universal gemm
+ * 2. Merge Batch with M and N dimension. It allows to increase compute in
+ *    case of small M and N. It also allows to vector load and store in case of
+ *    K = 1, C = 1 and HHWGC layout.
```
Comment (aosewski) [inline include/ck/tensor_operation/operator_transform/transform_conv_bwd_weight_to_gemm_v2.hpp]: Spelling:
```suggestion
 *    K = 1, C = 1 and NHWGC layout.
```

File: include/ck/tensor_operation/operator_transform/transform_conv_bwd_weight_to_gemm_v2.hpp [hunk 5]
```diff
@@ -0,0 +1,640 @@
...
+                       const index_t Y,
+                       const index_t X,
+                       const index_t C,
+                       const std::array<index_t, NDimSpatial + 3>& weights_strides)
+    {
+        const auto CStride     = Number<1>{};
+        const auto KStride     = weights_strides[1];
+        const auto XStride     = weights_strides[4];
+        const auto BatchStride = weights_strides[0];
+        // Add NumBatchToMerge for Batch+M dimension and, 1 for placehord for Batch+N dimension
```
Comment (aosewski) [inline include/ck/tensor_operation/operator_transform/transform_conv_bwd_weight_to_gemm_v2.hpp]:
```suggestion
        // Add NumBatchToMerge for Batch+M dimension and, 1 as a placeholder for Batch+N dimension
```

File: include/ck/tensor_operation/gpu/device/impl/device_grouped_conv_bwd_weight_two_stage_xdl_cshuffle.hpp [hunk 6]
```diff
@@ -183,162 +232,180 @@ struct DeviceGroupedConvBwdWeightTwoStage_Xdl_CShuffle
...
+                                                                         params,
+                                                                         batch)[I2];
     }
 
     using ABCGridDescs = decltype(GetABCGridDesc<NDimSpatial>());
 
-    using AGridDesc_K0_M_K1 = remove_cvref_t<decltype(ABCGridDescs{}[I0])>;
-    using BGridDesc_K0_N_K1 = remove_cvref_t<decltype(ABCGridDescs{}[I1])>;
+    using AGridDesc_K0_M_K1 = remove_cvref_t<decltype((ABCGridDescs{}[I0]))>;
+    using BGridDesc_K0_N_K1 = remove_cvref_t<decltype((ABCGridDescs{}[I1]))>;
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/device/impl/device_grouped_conv_bwd_weight_two_stage_xdl_cshuffle.hpp]: Are those double braces necessary?

File: include/ck/tensor_operation/gpu/grid/gridwise_gemm_xdl_cshuffle_bwd_weight_v3.hpp [hunk 7]
```diff
@@ -0,0 +1,1370 @@
...
+            make_tuple(Sequence<0>{}, Sequence<1>{}),
+            make_tuple(Sequence<0, 1>{}, Sequence<2, 3>{}));
+
+        return c_grid_desc_mblock_mperblock_nblock_nperblock;
+    }
+
+    // return block_id to C matrix tile idx (m0, n0) mapping
+    // if arch = gfx942
+    using Block2CTileMap = BlockToCTileMap_Grouped_M00_N0_M01Adapt<8, MPerBlock, NPerBlock>;
+    // using Block2CTileMap = BlockToCTileMap_3DGrid_KSplit<MPerBlock, NPerBlock>;
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/grid/gridwise_gemm_xdl_cshuffle_bwd_weight_v3.hpp]: Leftover comments.

**Review by aosewski — DISMISSED** [review]
Have just few minor comments. In general great work!

File: library/src/tensor_operation_instance/gpu/grouped_conv2d_bwd_weight/xdl/device_grouped_conv2d_bwd_weight_two_stage_xdl_nhwgc_gkyxc_nhwgk_f16_pipev2_instance.cpp (hunk 2 above)
Comment (bartekxk) [reply library/src/tensor_operation_instance/gpu/grouped_conv2d_bwd_weight/xdl/device_grouped_conv2d_bwd_weight_two_stage_xdl_nhwgc_gkyxc_nhwgk_f16_pipev2_instance.cpp:42]: Not applicable for theses cases. We should investigate in the future performance difference using this specialization.

**Review by aosewski — APPROVED** [review]

**Pull request merged**